File: R/regions.R
#' Cells whose centres lie inside a polygon
#'
#' Returns A5 cells at `resolution` whose centres fall inside the polygon.
#' Multi-feature inputs (a `MULTIPOLYGON`, an `sfc` of multiple polygons,
#' or a `POLYGON` with holes) are handled natively: per polygon part, the
#' outer ring and its holes are converted together with hole interiors
#' excluded, then the results are unioned across parts. The final cell set
#' is compacted; use [a5_uncompact()] to expand to a uniform-resolution grid.
#'
#' @param x A polygon-like geometry. One of:
#'   - Any geometry handleable by [wk::wk_handle()] (e.g. [wk::wkt()],
#'     [wk::wkb()], [wk::rct()], `sf`, `sfc`) containing one or more
#'     `POLYGON` / `MULTIPOLYGON` features.
#'   - A `SpatVector` of polygons (requires the `terra` package).
#'   - A two-column numeric matrix (`cbind(lon, lat)`) of vertices,
#'     interpreted as a single outer ring.
#'   - A `data.frame` with columns `lon` and `lat`, interpreted as a
#'     single outer ring.
#' @param resolution Integer scalar target resolution (0-30).
#'
#' @returns An [a5_cell] vector at or coarser than `resolution`.
#'
#' @details
#' Membership is determined by **centre-point containment**: a cell is
#' included iff its centroid lies inside the polygon, with hole rings
#' properly subtracted. This is distinct from [a5_grid()]'s
#' boundary-intersection semantics; for the same polygon the two
#' functions can return slightly different cell sets near the boundary.
#'
#' Coordinates must be WGS 84 longitude/latitude in degrees. Rings are
#' closed automatically; a trailing duplicate vertex is dropped if
#' present.
#'
#' Where no A5 cell centroids at the specified `resolution` fall within
#' the geometry, an empty `a5_cell` vector is returned.
#'
#' Matrix and `data.frame` inputs are treated as a single ring; for
#' multi-feature data or polygons with holes, pass an `sf`, `sfc`, wk,
#' or `SpatVector` geometry instead.
#'
#' @seealso [a5_linestring_to_cells()], [a5_uncompact()].
#' @export
#' @examples
#' poly <- wk::wkt(
#'   "POLYGON ((-3.3 55.9, -3.1 55.9, -3.1 56, -3.3 56, -3.3 55.9))"
#' )
#' cells <- a5_polygon_to_cells(poly, resolution = 8)
#' length(cells)
a5_polygon_to_cells <- function(x, resolution) {
  resolution <- vctrs::vec_cast(resolution, integer())
  check_resolution(resolution)
  vctrs::vec_assert(resolution, size = 1L)

  bundle <- prepare_polygon_input(x)

  cells_from_rs(a5_polygon_to_cells_rs(
    bundle$lon,
    bundle$lat,
    bundle$offsets,
    bundle$part_id,
    bundle$is_outer,
    resolution
  ))
}

#' Cells traced by a great-circle linestring
#'
#' Returns the A5 cells at `resolution` whose pentagons are intersected by
#' the great-circle polyline connecting the supplied waypoints. Output is
#' uncompacted, in discovery order along the path, with duplicates removed
#' first-seen. Multi-feature inputs (a `MULTILINESTRING` or an `sfc` of
#' multiple linestrings) are handled natively: per-feature cell sequences
#' are concatenated in feature order with first-seen deduplication.
#'
#' @param x A linestring-like geometry. One of:
#'   - Any geometry handleable by [wk::wk_handle()] containing one or
#'     more `LINESTRING` / `MULTILINESTRING` features.
#'   - A `SpatVector` of linestrings (requires the `terra` package).
#'   - A two-column numeric matrix (`cbind(lon, lat)`) of waypoints.
#'   - A `data.frame` with columns `lon` and `lat`.
#' @param resolution Integer scalar target resolution (0--30).
#'
#' @returns An [a5_cell] vector at `resolution`.
#'
#' @details
#' Consecutive waypoints are connected by great-circle arcs (not rhumb
#' lines or planar segments), so antimeridian-crossing paths work
#' transparently when written in unwrapped lon/lat.
#'
#' Matrix and `data.frame` inputs are treated as a single linestring;
#' for multi-feature data, pass an `sf`, `sfc`, wk, or `SpatVector`
#' geometry instead.
#'
#' @seealso [a5_polygon_to_cells()], [a5_grid_disk()].
#' @export
#' @examples
#' line <- wk::wkt("LINESTRING (2.35 48.86, -0.13 51.51)")
#' cells <- a5_linestring_to_cells(line, resolution = 6)
#' length(cells)
a5_linestring_to_cells <- function(x, resolution) {
  resolution <- vctrs::vec_cast(resolution, integer())
  check_resolution(resolution)
  vctrs::vec_assert(resolution, size = 1L)

  bundle <- prepare_linestring_input(x)

  cells_from_rs(a5_linestring_to_cells_rs(
    bundle$lon,
    bundle$lat,
    bundle$offsets,
    resolution
  ))
}

# -- input preparation ---------------------------------------------------------

#' Flatten polygon input to per-ring (lon, lat, offsets, part_id, is_outer)
#'
#' Returns `list(lon, lat, offsets, part_id, is_outer)`. `offsets` is
#' cumulative; ring `i` occupies `offsets[i] + 1` to `offsets[i + 1]`.
#' `part_id[i]` groups rings by polygon part. `is_outer[i]` is `1L` for
#' an outer ring and `0L` for a hole ring.
#' @noRd
prepare_polygon_input <- function(x, call = rlang::caller_env()) {
  x <- maybe_coerce_spatvector(x, call = call)
  bundle <- if (is.matrix(x)) {
    coords_from_matrix(x, call = call)
  } else if (is.data.frame(x) && !inherits(x, "sf")) {
    coords_from_data_frame(x, call = call)
  } else {
    coords_from_wk_polygons(x, call = call)
  }

  if (length(bundle$offsets) < 2L) {
    cli::cli_abort("{.arg x} contains no polygon rings.", call = call)
  }
  validate_feature_sizes(bundle, expected = "polygon", call = call)
  if (anyNA(bundle$lon) || anyNA(bundle$lat)) {
    cli::cli_abort(
      "{.arg x} must not contain {.code NA} coordinates.",
      call = call
    )
  }
  bundle
}

#' Flatten linestring input to (lon, lat, offsets)
#' @noRd
prepare_linestring_input <- function(x, call = rlang::caller_env()) {
  x <- maybe_coerce_spatvector(x, call = call)
  bundle <- if (is.matrix(x)) {
    coords_from_matrix(x, call = call)
  } else if (is.data.frame(x) && !inherits(x, "sf")) {
    coords_from_data_frame(x, call = call)
  } else {
    coords_from_wk_linestrings(x, call = call)
  }

  if (length(bundle$offsets) < 2L) {
    cli::cli_abort("{.arg x} contains no linestrings.", call = call)
  }
  validate_feature_sizes(bundle, expected = "linestring", call = call)
  if (anyNA(bundle$lon) || anyNA(bundle$lat)) {
    cli::cli_abort(
      "{.arg x} must not contain {.code NA} coordinates.",
      call = call
    )
  }
  bundle
}

#' @noRd
coords_from_matrix <- function(x, call = rlang::caller_env()) {
  if (ncol(x) != 2L) {
    cli::cli_abort(
      "Matrix input must have exactly 2 columns ({.code lon}, {.code lat}); \\
       got {ncol(x)}.",
      call = call
    )
  }
  if (!is.numeric(x)) {
    cli::cli_abort("Matrix input must be numeric.", call = call)
  }
  lon <- as.numeric(x[, 1L])
  lat <- as.numeric(x[, 2L])
  list(
    lon = lon,
    lat = lat,
    offsets = c(0L, length(lon)),
    part_id = 1L,
    is_outer = 1L
  )
}

#' @noRd
coords_from_data_frame <- function(x, call = rlang::caller_env()) {
  if (!all(c("lon", "lat") %in% names(x))) {
    cli::cli_abort(
      "Data frame input must have columns {.code lon} and {.code lat}.",
      call = call
    )
  }
  lon <- as.numeric(x$lon)
  lat <- as.numeric(x$lat)
  list(
    lon = lon,
    lat = lat,
    offsets = c(0L, length(lon)),
    part_id = 1L,
    is_outer = 1L
  )
}

#' @noRd
coords_from_wk_polygons <- function(x, call = rlang::caller_env()) {
  meta <- safe_wk_meta(x, call = call)

  # wk_meta() geometry_type codes: 3 = POLYGON, 6 = MULTIPOLYGON.
  allowed <- c(POLYGON = 3L, MULTIPOLYGON = 6L)
  bad <- !is.na(meta$geometry_type) & !(meta$geometry_type %in% allowed)
  if (any(bad)) {
    cli::cli_abort(
      "{.arg x} must contain only {paste(names(allowed), collapse = ' or ')} geometries.",
      call = call
    )
  }
  if (any(meta$is_empty)) {
    cli::cli_abort("{.arg x} contains empty geometries.", call = call)
  }

  coords <- wk::wk_coords(x)
  if (nrow(coords) == 0L) {
    return(list(
      lon = numeric(),
      lat = numeric(),
      offsets = 0L,
      part_id = integer(),
      is_outer = integer()
    ))
  }

  # Ring grouping (feature_id, part_id, ring_id) — first-seen order.
  ring_key <- paste(
    coords$feature_id,
    coords$part_id,
    coords$ring_id,
    sep = "/"
  )
  ring_group <- match(ring_key, unique(ring_key))

  # Polygon-part grouping (feature_id, part_id) — first-seen order.
  part_key <- paste(coords$feature_id, coords$part_id, sep = "/")
  row_part_id <- match(part_key, unique(part_key))

  rle_rings <- rle(ring_group)
  offsets <- c(0L, cumsum(rle_rings$lengths))
  ring_starts <- offsets[-length(offsets)] + 1L

  ring_part_id <- row_part_id[ring_starts]
  # Within a polygon part, the first ring (smallest ring_id) is the outer
  # and any later rings are holes. wk emits rings in outer-first order, so
  # the first occurrence of each part_id marks the outer ring.
  ring_is_outer <- as.integer(!duplicated(ring_part_id))

  # Closed rings (trailing duplicate vertex) are passed through as-is;
  # `a5::polygon_to_cells` strips the closing vertex natively.
  list(
    lon = as.numeric(coords$x),
    lat = as.numeric(coords$y),
    offsets = as.integer(offsets),
    part_id = as.integer(ring_part_id),
    is_outer = as.integer(ring_is_outer)
  )
}

#' @noRd
coords_from_wk_linestrings <- function(x, call = rlang::caller_env()) {
  meta <- safe_wk_meta(x, call = call)

  # wk_meta() geometry_type codes: 2 = LINESTRING, 5 = MULTILINESTRING.
  allowed <- c(LINESTRING = 2L, MULTILINESTRING = 5L)
  bad <- !is.na(meta$geometry_type) & !(meta$geometry_type %in% allowed)
  if (any(bad)) {
    cli::cli_abort(
      "{.arg x} must contain only {paste(names(allowed), collapse = ' or ')} geometries.",
      call = call
    )
  }
  if (any(meta$is_empty)) {
    cli::cli_abort("{.arg x} contains empty geometries.", call = call)
  }

  coords <- wk::wk_coords(x)
  if (nrow(coords) == 0L) {
    return(list(lon = numeric(), lat = numeric(), offsets = 0L))
  }

  # Linestrings: one ring_id per linestring (within a feature/part).
  line_key <- paste(
    coords$feature_id,
    coords$part_id,
    coords$ring_id,
    sep = "/"
  )
  line_group <- match(line_key, unique(line_key))
  rle_lines <- rle(line_group)
  offsets <- c(0L, cumsum(rle_lines$lengths))

  list(
    lon = as.numeric(coords$x),
    lat = as.numeric(coords$y),
    offsets = as.integer(offsets)
  )
}

#' Coerce a terra `SpatVector` to a wk-handleable WKT vector
#'
#' terra does not register `wk_handle()` methods, so we route SpatVector
#' inputs through `terra::geom(x, wkt = TRUE)` and wrap the result in a
#' `wk_wkt` vector. `terra` is in `Suggests`; this path errors with a
#' clear install hint if the package is missing.
#' @noRd
maybe_coerce_spatvector <- function(x, call = rlang::caller_env()) {
  if (!inherits(x, "SpatVector")) {
    return(x)
  }
  rlang::check_installed("terra", reason = "to convert terra SpatVector input.")
  wk::wkt(terra::geom(x, wkt = TRUE))
}

#' @noRd
safe_wk_meta <- function(x, call = rlang::caller_env()) {
  tryCatch(
    wk::wk_meta(x),
    error = function(e) {
      cli::cli_abort(
        c(
          "Could not interpret {.arg x} as a geometry.",
          i = "Pass a matrix, data frame with {.code lon}/{.code lat} \\
               columns, or a {.fn wk::wk_handle}-able object."
        ),
        parent = e,
        call = call
      )
    }
  )
}

#' @noRd
validate_feature_sizes <- function(
  bundle,
  expected,
  call = rlang::caller_env()
) {
  sizes <- diff(bundle$offsets)
  min_size <- switch(expected, polygon = 3L, linestring = 2L)
  short <- which(sizes < min_size)
  if (length(short) > 0L) {
    cli::cli_abort(
      "Each {expected} ring must have at least {min_size} vertices; \\
       ring {short[1]} has {sizes[short[1]]}.",
      call = call
    )
  }
  invisible(bundle)
}

File: src/rust/src/regions.rs
use extendr_api::prelude::*;
use rayon::prelude::*;
use std::collections::HashSet;
use std::result::Result as StdResult;

use crate::cell_raw::u64s_to_raw8_list;
use crate::threading::{get_num_threads, maybe_par};

/// Build per-ring `Vec<a5::LonLat>`s from flat coordinate slices and offsets.
fn split_rings(lon: &Doubles, lat: &Doubles, offsets: &Integers) -> Vec<Vec<a5::LonLat>> {
    let n_rings = offsets.len().saturating_sub(1);
    let mut out: Vec<Vec<a5::LonLat>> = Vec::with_capacity(n_rings);
    for i in 0..n_rings {
        let start = offsets[i].inner() as usize;
        let end = offsets[i + 1].inner() as usize;
        let mut ring: Vec<a5::LonLat> = Vec::with_capacity(end.saturating_sub(start));
        for j in start..end {
            ring.push(a5::LonLat::new(lon[j].inner(), lat[j].inner()));
        }
        out.push(ring);
    }
    out
}

/// Per-part processing: outer cells minus union of hole cells, all at `resolution`.
fn process_part(
    p: i32,
    rings: &[Vec<a5::LonLat>],
    part_id: &[i32],
    is_outer: &[i32],
    resolution: i32,
) -> StdResult<Vec<u64>, String> {
    let mut outer_ring_idx: Option<usize> = None;
    let mut hole_ring_indices: Vec<usize> = Vec::new();
    for (i, &pid) in part_id.iter().enumerate() {
        if pid != p {
            continue;
        }
        if is_outer[i] != 0 {
            outer_ring_idx = Some(i);
        } else {
            hole_ring_indices.push(i);
        }
    }
    let oi = match outer_ring_idx {
        Some(i) => i,
        None => return Ok(Vec::new()),
    };

    let outer_compacted = a5::polygon_to_cells(&rings[oi], resolution)?;
    let outer_uncompacted = a5::uncompact(&outer_compacted, resolution)?;
    let mut cells: HashSet<u64> = outer_uncompacted.into_iter().collect();

    for &hi in &hole_ring_indices {
        let hole_compacted = a5::polygon_to_cells(&rings[hi], resolution)?;
        let hole_uncompacted = a5::uncompact(&hole_compacted, resolution)?;
        for h in hole_uncompacted {
            cells.remove(&h);
        }
    }

    Ok(cells.into_iter().collect())
}

/// Convert one or more polygon parts (with optional holes) to A5 cells.
///
/// `lon`, `lat` are flat coordinate vectors; `offsets` is cumulative
/// (length `n_rings + 1`) so ring `i` is `lon[offsets[i]..offsets[i+1]]`.
/// `part_id` (length `n_rings`) groups rings by polygon part. `is_outer`
/// (length `n_rings`, 1 = outer / 0 = hole) classifies each ring.
///
/// For each polygon part, the outer ring is converted to cells, the
/// uncompacted cells inside any hole rings are subtracted, and the
/// remaining cells are unioned across parts. The result is recompacted
/// at the end. A single-outer-no-holes input takes a fast path that
/// passes the upstream compacted result through unchanged.
///
/// @noRd
/// @keywords internal
#[extendr]
fn a5_polygon_to_cells_rs(
    lon: Doubles,
    lat: Doubles,
    offsets: Integers,
    part_id: Integers,
    is_outer: Integers,
    resolution: i32,
) -> List {
    let rings = split_rings(&lon, &lat, &offsets);
    let n_rings = rings.len();

    if n_rings == 0 {
        return u64s_to_raw8_list(vec![]);
    }

    let part_id_vec: Vec<i32> = part_id.iter().map(|p| p.inner()).collect();
    let is_outer_vec: Vec<i32> = is_outer.iter().map(|b| b.inner()).collect();

    let max_part = part_id_vec.iter().copied().max().unwrap_or(0);
    if max_part <= 0 {
        return u64s_to_raw8_list(vec![]);
    }

    // Fast path: a single outer ring with no holes (the common case for
    // matrix / data.frame input and for an `sfc` of one POLYGON).
    let total_outer = is_outer_vec.iter().filter(|&&v| v != 0).count();
    if max_part == 1 && n_rings == 1 && total_outer == 1 {
        return match a5::polygon_to_cells(&rings[0], resolution) {
            Ok(cells) => {
                let out: Vec<Option<u64>> = cells.into_iter().map(Some).collect();
                u64s_to_raw8_list(out)
            }
            Err(e) => throw_r_error(format!("polygon_to_cells failed: {}", e)),
        };
    }

    let parts: Vec<i32> = (1..=max_part).collect();
    let per_part: Vec<StdResult<Vec<u64>, String>> = if get_num_threads() <= 1 {
        parts
            .iter()
            .map(|&p| process_part(p, &rings, &part_id_vec, &is_outer_vec, resolution))
            .collect()
    } else {
        maybe_par(|| {
            parts
                .par_iter()
                .map(|&p| process_part(p, &rings, &part_id_vec, &is_outer_vec, resolution))
                .collect()
        })
    };

    let mut union: HashSet<u64> = HashSet::new();
    for r in per_part {
        let part_cells = match r {
            Ok(c) => c,
            Err(e) => throw_r_error(format!("polygon_to_cells failed: {}", e)),
        };
        union.extend(part_cells);
    }

    let mut combined: Vec<u64> = union.into_iter().collect();
    combined.sort_unstable();
    let recompacted = match a5::compact(&combined) {
        Ok(c) => c,
        Err(e) => throw_r_error(format!("compact failed: {}", e)),
    };

    let out: Vec<Option<u64>> = recompacted.into_iter().map(Some).collect();
    u64s_to_raw8_list(out)
}

/// Convert one or more linestring waypoint sequences to A5 cells.
///
/// Per-feature `line_string_to_cells` results are concatenated in feature
/// order with first-seen deduplication, preserving discovery order along
/// the path.
///
/// @noRd
/// @keywords internal
#[extendr]
fn a5_linestring_to_cells_rs(
    lon: Doubles,
    lat: Doubles,
    offsets: Integers,
    resolution: i32,
) -> List {
    let features = split_rings(&lon, &lat, &offsets);

    if features.is_empty() {
        return u64s_to_raw8_list(vec![]);
    }

    let per_feature: Vec<StdResult<Vec<u64>, String>> =
        if get_num_threads() <= 1 || features.len() == 1 {
            features
                .iter()
                .map(|w| a5::line_string_to_cells(w, resolution))
                .collect()
        } else {
            maybe_par(|| {
                features
                    .par_iter()
                    .map(|w| a5::line_string_to_cells(w, resolution))
                    .collect()
            })
        };

    let mut seen: HashSet<u64> = HashSet::new();
    let mut result: Vec<u64> = Vec::new();
    for r in per_feature {
        let cells = match r {
            Ok(c) => c,
            Err(e) => throw_r_error(format!("line_string_to_cells failed: {}", e)),
        };
        for c in cells {
            if seen.insert(c) {
                result.push(c);
            }
        }
    }

    let out: Vec<Option<u64>> = result.into_iter().map(Some).collect();
    u64s_to_raw8_list(out)
}

extendr_module! {
    mod regions;
    fn a5_polygon_to_cells_rs;
    fn a5_linestring_to_cells_rs;
}
